proto/api_v3/query_service.proto

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2021 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax="proto3";
+
+package jaeger.api_v3;
+
+import "opentelemetry/proto/trace/v1/trace.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/duration.proto";
+
+option go_package = "api_v3";
+option java_package = "io.jaegertracing.api_v3";
+
+// Request object to get a trace.
+message GetTraceRequest {
+  string trace_id = 1;
+}
+
+// A single response chunk holds a single trace.
+// Note that a subsequent request to get a trace might return a different result
+// because spans are reported in asynchronous chunks.
+message SpansResponseChunk {
+  // OTLP resource spans
+  // see https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto
+  repeated opentelemetry.proto.trace.v1.ResourceSpans resource_spans = 1;
+}
+
+// Query parameters to find traces.
+message TraceQueryParameters {
+  string service_name = 1;
+  string operation_name = 2;
+  map<string, string> attributes = 3;
+  google.protobuf.Timestamp start_time_min = 4;
+  google.protobuf.Timestamp start_time_max = 5;
+  google.protobuf.Duration duration_min = 6;
+  google.protobuf.Duration duration_max = 7;
+  int32 num_traces = 8;
+}
+
+// Request object to search traces.
+message FindTracesRequest {
+  TraceQueryParameters query = 1;
+}
+
+// Request object to get service names.
+message GetServicesRequest {}
+
+// Response object to get service names.
+message GetServicesResponse {
+  repeated string services = 1;
+}
+
+// Request object to search for operation names.
+message GetOperationsRequest {
+  string service = 1;
+  string span_kind = 2;
+}
+
+// Operation encapsulates information about operation.
+message Operation {
+  string name = 1;
+  string span_kind = 2;
+}
+
+// Response object to get operation names.
+message GetOperationsResponse {
+  repeated Operation operations = 1;
+}
+
+service QueryService {
+  // GetTraces returns single trace.
+  rpc GetTrace(GetTraceRequest) returns (stream SpansResponseChunk) {}
+
+  // GetTraces searches for traces.
+  rpc GetTraces(FindTracesRequest) returns (stream SpansResponseChunk) {}
```

Comment:
@yurishkuro do you remember why streaming was used? An alternative would be to return a list of chunks. Also we could remove chunk to `Trace` which is more idiomatic.

One issue with the streaming and grpc-gateway is that it wraps the response into result - e.g. `{result: {resource_spans: ...}}` see https://github.com/jaegertracing/jaeger/pull/3086/files#diff-1429f7cc5a76981a44799039e43d3bc7372808373b9c2b97a333c7dcf650b00aR72

Reply:
Streaming - because large result sets are difficult to transmit as one response.

Returning chunks manually means implementing some kind of pagination API, which would require support in the storage.

Reply:
one way or another to fully support streaming the storage API would have to change anyways.

Reply:
storage API already has `FindTraceIDs`, which allows the query service to load traces one by one